src/Parsers/TokenIterator.h

```diff
@@ -20,34 +20,18 @@ class Tokens
 {
 private:
     std::vector<Token> data;
-    Lexer lexer;
+    std::size_t last_accessed_index = 0;
 
 public:
-    Tokens(const char * begin, const char * end, size_t max_query_size = 0) : lexer(begin, end, max_query_size) {}
+    Tokens(const char * begin, const char * end, size_t max_query_size = 0);
 
-    const Token & operator[] (size_t index)
+    ALWAYS_INLINE inline const Token & operator[](size_t index)
     {
-        while (true)
-        {
-            if (index < data.size())
-                return data[index];
-
-            if (!data.empty() && data.back().isEnd())
-                return data.back();
-
-            Token token = lexer.nextToken();
-
-            if (token.isSignificant())
-                data.emplace_back(token);
-        }
+        last_accessed_index = std::max(last_accessed_index, index);
+        return data[index];
```

Comment:
It seems like now we don't check whether the index is out of range

Reply:
As it's something that shouldn't happen I'll add an chassert to verify it in debug builds.